Admin UI: Login improvements

zenflow opened the pull request:
Closes #2656 with this change:
- After user signs in, redirect them where they were trying to go (6fc6331)

Closes #2655 with these changes:
 - Show AccessDeniedPage when access is denied & user is already signed in (23a68cf)
 - Handle post-signin redirects also for unauthorized users (5e6fa04) 

I updated the cypress tests (for test-projects/login) in these regards (6fc6331).

I also noticed that a lot of these tests were disabled! I'm not sure why that is, but I re-enabled them (4890f5c) and, since #2656 is resolved here, all these tests including ["Redirects to requested page after login"](https://github.com/keystonejs/keystone/blob/d138736db184c5884171c7a65e43377f248046b5/test-projects/login/cypress/integration/login_spec.js#L182-L198) are passing.

### Update

Closes #2658 with these changes:
- auth-passport: Added getAdminMeta method required by Admin UI (382bfc9)
- auth-passport: added resolveState config to compute oath `state` param (e76f9a3)
See "State" under https://developers.google.com/identity/protocols/oauth2/web-server#creatingclient
- auth-passport: Make loginPath a public member (10aa5ee)
- app-admin-ui: Support auth strategies that have `loginPath` property, closes #2658 (f06410a)

zenflow commented:
BTW Here is what the Access Denied screen looks like:
![image](https://user-images.githubusercontent.com/3198597/78322667-653a5d00-753d-11ea-88d7-3f704690de84.png)

"Go Home" links to (`/`)

Vultraz commented:
I wonder, should the Access Denied page perhaps look more like my new Signout page (#2471)? 🤔

zenflow commented:
@Vultraz Yeah, I believe it should be made to look like your new Signout page. I guess we should do this when our changes are merged in Git? Either this branch is merged first and you get to apply your Logout changes to AccessDenied, or your branch is merged first and I get the honor. I think it should be simple either way because AccessDenied uses the exact same design & layout as Signout.

zenflow commented:
See "Update" in initial PR comment:
>Closes #2658 with these changes:
>- auth-passport: Added getAdminMeta method required by Admin UI (382bfc9)
>- auth-passport: added resolveState config to compute oath state param (e76f9a3)
>See "State" under https://developers.google.com/identity/protocols/oauth2/web-server#creatingclient
>- auth-passport: Make loginPath a public member (10aa5ee)
>- app-admin-ui: Support auth strategies that have loginPath property, closes #2658 (f06410a)

Vultraz commented:
Yeah, I agree. We'll just see which gets merged first.

BTW, about your changesets. You don't need to add one for each change. You can just add a single changeset that covers all the packages you updated and a list/description of each of your changes.

Vultraz commented on packages/app-admin-ui/client/pages/AccessDeniedPage.js:
```diff
@@ -0,0 +1,71 @@
+});
+
+const AccessDeniedPage = () => {
+  const {} = useAdminMeta();
```
No need for this since you're not using anything from the admin metadata.

Vultraz commented on packages/app-admin-ui/index.js:
```diff
@@ -25,8 +26,8 @@ class AdminUIApp {
 
-    if (authStrategy && authStrategy.authType !== 'password') {
-      throw new Error('Keystone 5 Admin currently only supports the `PasswordAuthStrategy`');
+    if (authStrategy && authStrategy.authType !== 'password' && !authStrategy.loginPath) {
```
Do you still need the check against the `password` auth type?

zenflow replied:
Yeah because the password auth strategy doesn't have a `.loginPath` property. We don't require a password auth to have a loginPath to redirect to because this package (app-admin-ui) is itself able to handle login flow for password auth. 

So the requirement for an auth strategy is to (1) be password type (login web flow can be handled by package) **or** (2) have a loginPath (which handles login web flow).

gautamsi commented:
I suggest that we create standalone components for auth buttons for social auth. Also it would be great if the social auth package (auth-passport or custom one) contribute to the hook for signin page login buttons

zenflow commented:
> Also it would be great if the social auth package (auth-passport or custom one) contribute to the hook for signin page login buttons

@gautamsi Correct me if I'm wrong: The Admin UI supports only one `authStrategy` and thus only one oauth (social signin) provider. If it's true then there is no use of a signin page because all paths lead to one, and we can simplify the flow by removing it. In other words: When a user goes to `/admin/signin` (or is redirected there), instead of showing a page with a "Sign in" button, we should just sign them in.

gautamsi commented:
@zenflow  yes it makes sense to redirect if the user is already signed in (when social auth was used)

What I meant that we can add buttons for social auth which would just be redirecting you to get social auth done. this button for each social auth type could be contributed from respective package rather than manually added by developer. 

I may have wrongly assumed that you are also putting a button for social auth on sign in page. if that is incorrect ignore my comments.

zenflow commented:
> I may have wrongly assumed that you are also putting a button for social auth on sign in page. if that is incorrect ignore my comments.

@gautamsi Yes, that was a wrong assumption. I didn't change the sign in page at all, and social sign in doesn't use the sign in page.

To be clear:
- With password auth, the `/admin/signin` route hosts the `Signin` page
- With auth types that define `loginPath` property (i.e. any passport auth), the `/admin/signin` route just redirects to the loginPath and *the Signin page is not used*.

zenflow commented:
@timleslie Any interest in getting this merged? I really think these are some very simple but very beneficial changes. I actually ended up not using Keystone, but I still think it would be really great for Keystone to have these features.

If there is interest in getting this merged, I will:
- Resolve the merge conflicts 
- Make the new Access Denied page look like the updated Signout page, as per https://github.com/keystonejs/keystone/pull/2657#issuecomment-609361673
- Anything else?

Vultraz commented:
I think it might be beneficial to split this into multiple PRs, one for each issue you address. Easier to review that way, and for the simpler stuff to get merged.

MadeByMike commented:
There is a lot of good stuff here but I only got half way through reviewing it last time. I notice a lot of conflicts now. @zenflow how much work would it be to make a few smaller PRs out fo this for each issue?

If not I might pull them out over time myself.

zenflow commented:
@MadeByMike @Vultraz Yeah, I'd be happy to contribute these changes in smaller PRs, since there is the interest in merging it

I'll start with the fix for #2656 which is a small change to the source.

Vultraz commented:
I see no reason why most if not all of these wouldn't be merged. 👍 

BTW, since you're touching this related area, there's something I've been pondering how best to solve. After signout, you can still technically go back to the admin UI. It doesn't *work*, but it would still be good not to be able to do that and redirect immediately to the Sign In page (or see the Access Denied page)... not entirely sure that's possible 🤔 

Alternatively, I could toss the Sign Out page entirely and have it immediately just show the sign-in page... Since there's a reload there, that would load up the public-facing entry point and (I think) you wouldn't be able to just go Back to the admin ui.

zenflow commented:
> After signout, you can still technically go back to the admin UI

@Vultraz Yes, I guess that's possible if there are any react-router links on the page. There is definitely some technical debt regarding that signout page and the public/private app versions (some of which I would need to deal with in my changes), and I don't really see any use for it anyways, so I'd like to remove it in the way you describe.

It would be nice for everyone to get it out of the way before making the PR for #2656.. 

Should I go ahead and make that the first PR?

/cc @MadeByMike

Vultraz commented:
What exactly are you planning on doing? I was imagining making a new `SignOutButton` component that handles the unauthenticate mutation (same as the signout page does now), along with a reload when the mutation completes (like the sign*in* page does now). And I was thinking it would be nice to have a "You have been signed out" message on the signin screen.

MadeByMike commented:
@zenflow I don't know exactly how you plan to split these. Can we get a checklist happening: 

- [ ] Redirect users after login #2656
- [ ] ...

zenflow commented:
> What exactly are you planning on doing?
@Vultraz Sorry I missed that part. Actually I assumed (foolish) that we could easily replace the `/admin/signout` page with a server-side express route that logs the user out and redirects the page to login, but I see that's not exactly the case since we need to log the user out via the graphql api.

Yeah, your way sounds slick. I can't say for sure if I'll get to it though. I would open an issue for it. (No really, I would open an issue for it but I'm not sure if y'all are using Github Issues for this type of thing or just bug tracking now)

zenflow commented:
@MadeByMike I was thinking roughly one PR per issue:
1. Redirect users after login #2656 (Note: This will handle both authorized and unauthorized users; There's no point of extending the problem of #2655.)
2. Access Denied screen #2655 (Assuming the issue is accepted and re-opened..) 
3. Support PassportAuthStrategy #2334

Yakoust commented:
Any update on this PR?
Espacially on `PassportAuthStrategy`?

Neonox31 commented:
I am also very interested in using passport auth strategy, any news on it ?

dabbid commented:
@MadeByMike @zenflow Has this PR been given up?

dimaip commented:
Any plans to get this PR in? I desperately need to be able to authenticate with JWT tokens for accessing admin UI.

bladey commented:
Hi @zenflow and other community members,

Thanks for taking the time to submit this PR and participating in the ongoing discussion regarding these login improvements.

These changes touch a lot of the system with a number of moving parts, it would be best at this stage to fork the repository and make the changes in a seperate repository.

As you might already know, **Keystone 5** has officially moved into active maintenance mode as we push towards the next major new version **Keystone Next**, so time will be limited to concentrate on this specific PR.

However we would welcome an updated PR in the Keystone 5 repository that is mergeable, please let us know!
